File: edukate-checker/src/main/java/io/github/sanyavertolet/edukate/checker/services/impl/SpringAiChatService.java
package io.github.sanyavertolet.edukate.checker.services.impl;

import io.github.sanyavertolet.edukate.checker.domain.RequestContext;
import io.github.sanyavertolet.edukate.checker.dtos.ModelResponse;
import io.github.sanyavertolet.edukate.checker.services.ChatService;
import lombok.AllArgsConstructor;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import org.springframework.ai.chat.client.ChatClient;
import org.springframework.ai.content.Media;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;
import reactor.core.scheduler.Schedulers;

import java.util.List;


@Slf4j
@Service
@Profile("!silent")
@AllArgsConstructor
public class SpringAiChatService implements ChatService {
    private final ChatClient chatClient;

    @Override
    public Mono<ModelResponse> makeRequest(@NonNull RequestContext context) {
        return Mono.fromCallable(() -> callModel(context))
                .doOnSuccess(_ -> log.debug("Successfully retrieved AI response"))
                .subscribeOn(Schedulers.boundedElastic());
    }

    private ModelResponse callModel(@NonNull RequestContext ctx) {
        return structuredCallModel(ctx.problemText(), ctx.problemImages(), ctx.submissionImages());
    }

    private ModelResponse structuredCallModel(String problemText, List<Media> problemMedia, List<Media> submissionMedia) {
        return chatClient.prompt()
                .system(s -> s.param("problemText", problemText))
                .user(u -> u
                        .text("Here are the images for you to check as well as the problem image.")
                        .media(problemMedia.toArray(new Media[0]))
                        .media(submissionMedia.toArray(new Media[0]))
                )
                .call()
                .entity(ModelResponse.class);
    }
}

File: edukate-checker/src/main/kotlin/io/github/sanyavertolet/edukate/checker/services/impl/SpringAiChatService.kt
package io.github.sanyavertolet.edukate.checker.services.impl

import io.github.sanyavertolet.edukate.checker.domain.RequestContext
import io.github.sanyavertolet.edukate.checker.dtos.ModelResponse
import io.github.sanyavertolet.edukate.checker.services.ChatService
import org.slf4j.LoggerFactory
import org.springframework.ai.chat.client.ChatClient
import org.springframework.ai.chat.client.advisor.SimpleLoggerAdvisor
import org.springframework.ai.chat.model.ChatModel
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono
import reactor.core.scheduler.Schedulers

@Service
@Profile("!silent")
class SpringAiChatService(
    chatModel: ChatModel,
    @param:Value($$"${checker.openai.system-prompt}") private val systemPromptTemplate: String,
) : ChatService {
    private val chatClient = ChatClient.builder(chatModel).defaultAdvisors(SimpleLoggerAdvisor()).build()

    override fun makeRequest(ctx: RequestContext): Mono<ModelResponse> =
        Mono.fromCallable { callModel(ctx) }
            .doOnSuccess { log.debug("Successfully retrieved AI response") }
            .subscribeOn(Schedulers.boundedElastic())

    private fun callModel(ctx: RequestContext): ModelResponse {
        val allMedia = (ctx.problemImages + ctx.submissionImages).toTypedArray()
        return requireNotNull(
            chatClient
                .prompt()
                .system { s -> s.text(systemPromptTemplate).param("problemText", ctx.problemText) }
                .user { u -> u.text("Here are the images for you to check as well as the problem image.").media(*allMedia) }
                .call()
                .entity(ModelResponse::class.java)
        ) {
            "AI returned null response"
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(SpringAiChatService::class.java)
    }
}
